src/firmware_main_loop.c

```diff
@@ -30,28 +35,72 @@ void firmware_main_loop(void)
     // This starts the async orientation screen workflow, which is processed by the loop below.
     orientation_screen();
 
-    while (1) {
-        screen_process();
-        /* And finally, run the high-level event processing. */
+    const uint8_t* hww_data = NULL;
+    uint8_t hww_frame[USB_REPORT_SIZE] = {0};
 
-        rust_workflow_spin();
-
-        if (usb_is_enabled()) {
-            rust_async_usb_spin();
+#if APP_U2F == 1
+    u2f_packet_init();
+    const uint8_t* u2f_data = NULL;
+    uint8_t u2f_frame[USB_REPORT_SIZE] = {0};
+#endif
 
-            /* First, process all the incoming USB traffic. */
-            usb_processing_process(usb_processing_hww());
+    while (1) {
+        // Do USB I/O
+        if (!hww_data) {
+            hww_data = queue_pull(queue_hww_queue());
+        }
 #if APP_U2F == 1
-            usb_processing_process(usb_processing_u2f());
+        // Generate timeout packets
+        uint32_t timeout_cid;
+        while (u2f_packet_timeout_get(&timeout_cid)) {
```

Comment:
We have to wait for a timeout? Why?

Reply:
This is a loop checking if any of the current u2f commands have timed out. It was moved from the usb_processing file because it called `send` and that function has been removed since all Io should be done from the main loop.

Reply:
In the original, `send()` was called inside the loop. I am not really following the logic, but wouldn't that mean that here we should turn the  `while` into a `if`, so if there are multiple timeouts, their packets all get processed properly? Or is the queue large enough so it doesn't matter?

Reply:
Yeah, queue is very large